File: algorithms/BestTimeToBuyAndSellStockII/BestTimeToBuyAndSellStockII.py
class Solution(object):
    def maxProfit(self, prices):
        n = len(prices)
        profit = 0
        for i in range(1, n):
            if prices[i] > prices[i - 1]:
                profit += prices[i] - prices[i - 1]
        return profit

class Solution(object):
    def maxProfit(self, prices):
        from sys import maxsize
        buy = -maxsize
        sell = 0
        for price in prices:
            buy = max(buy, sell - price)
            sell = max(sell, buy + price)
        return sell

File: algorithms/BestTimeToBuyAndSellStockII/BestTimeToBuyAndSellStockII.cpp
/*
 * time: O(n), space: O(1)
 */
#include <iostream>
#include <vector>
using namespace std;

class Solution{
public:
    int maxProfit(vector<int>& prices){
        int n = prices.size();
        if(n <= 1)
            return 0;
        int maxProfit = 0;
        for(int i = 1; i < n; i++)
            maxProfit += max(0, prices[i] - prices[i - 1]);

        return maxProfit;
    }
};

int main(){
    int priceArray[] = {1, 4, 2};
    vector<int> prices = vector<int>(priceArray, priceArray + sizeof(priceArray) / sizeof(int));

    Solution solution;
    int res = solution.maxProfit(prices);
    cout << res << endl;

    return 0;
}


